From c1c1eb1b2d1d034dfe351ea51efd25ef502db25e Mon Sep 17 00:00:00 2001
From: Hans de Goede <hdegoede@redhat.com>
Date: Sun, 10 Oct 2021 20:56:57 +0200
Subject: [PATCH] ACPI: delay enumeration of devices with a _DEP pointing to an
 INT3472 device

The clk and regulator frameworks expect clk/regulator consumer-devices
to have info about the consumed clks/regulators described in the device's
fw_node.

To work around cases where this info is not present in the firmware tables,
which is often the case on x86/ACPI devices, both frameworks allow the
provider-driver to attach info about consumers to the clks/regulators
when registering these.

This causes problems with the probe ordering wrt drivers for consumers
of these clks/regulators. Since the lookups are only registered when the
provider-driver binds, trying to get these clks/regulators before then
results in a -ENOENT error for clks and a dummy regulator for regulators.

One case where we hit this issue is camera sensors such as e.g. the OV8865
sensor found on the Microsoft Surface Go. The sensor uses clks, regulators
and GPIOs provided by a TPS68470 PMIC which is described in an INT3472
ACPI device. There is special platform code handling this and setting
platform_data with the necessary consumer info on the MFD cells
instantiated for the PMIC under: drivers/platform/x86/intel/int3472.

For this to work properly the ov8865 driver must not bind to the I2C-client
for the OV8865 sensor until after the TPS68470 PMIC gpio, regulator and
clk MFD cells have all been fully setup.

The OV8865 on the Microsoft Surface Go is just one example, all X86
devices using the Intel IPU3 camera block found on recent Intel SoCs
have similar issues where there is an INT3472 HID ACPI-device, which
describes the clks and regulators, and the driver for this INT3472 device
must be fully initialized before the sensor driver (any sensor driver)
binds for things to work properly.

On these devices the ACPI nodes describing the sensors all have a _DEP
dependency on the matching INT3472 ACPI device (there is one per sensor).

This allows solving the probe-ordering problem by delaying the enumeration
(instantiation of the I2C-client in the ov8865 example) of ACPI-devices
which have a _DEP dependency on an INT3472 device.

The new acpi_dev_ready_for_enumeration() helper used for this is also
exported because for devices, which have the enumeration_by_parent flag
set, the parent-driver will do its own scan of child ACPI devices and
it will try to enumerate those during its probe(). Code doing this such
as e.g. the i2c-core-acpi.c code must call this new helper to ensure
that it too delays the enumeration until all the _DEP dependencies are
met on devices which have the new honor_deps flag set.

Signed-off-by: Hans de Goede <hdegoede@redhat.com>
Patchset: cameras
---
 drivers/acpi/scan.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/drivers/acpi/scan.c b/drivers/acpi/scan.c
index b78f6be2f..801171e2a 100644
--- a/drivers/acpi/scan.c
+++ b/drivers/acpi/scan.c
@@ -2206,6 +2206,9 @@ static acpi_status acpi_bus_check_add_2(acpi_handle handle, u32 lvl_not_used,
 
 static void acpi_default_enumeration(struct acpi_device *device)
 {
+	if (!acpi_dev_ready_for_enumeration(device))
+		return;
+
 	/*
 	 * Do not enumerate devices with enumeration_by_parent flag set as
 	 * they will be enumerated by their respective parents.
-- 
2.53.0

From 4b553e59c7d3ff31429e4dd08a0460b11bab94dd Mon Sep 17 00:00:00 2001
From: zouxiaoh <xiaohong.zou@intel.com>
Date: Fri, 25 Jun 2021 08:52:59 +0800
Subject: [PATCH] iommu: intel-ipu: use IOMMU passthrough mode for Intel IPUs

Intel IPU(Image Processing Unit) has its own (IO)MMU hardware,
The IPU driver allocates its own page table that is not mapped
via the DMA, and thus the Intel IOMMU driver blocks access giving
this error: DMAR: DRHD: handling fault status reg 3 DMAR:
[DMA Read] Request device [00:05.0] PASID ffffffff
fault addr 76406000 [fault reason 06] PTE Read access is not set
As IPU is not an external facing device which is not risky, so use
IOMMU passthrough mode for Intel IPUs.

Change-Id: I6dcccdadac308cf42e20a18e1b593381391e3e6b
Depends-On: Iacd67578e8c6a9b9ac73285f52b4081b72fb68a6
Tracked-On: #JIITL8-411
Signed-off-by: Bingbu Cao <bingbu.cao@intel.com>
Signed-off-by: zouxiaoh <xiaohong.zou@intel.com>
Signed-off-by: Xu Chongyang <chongyang.xu@intel.com>
Patchset: cameras
---
 drivers/iommu/intel/iommu.c | 30 ++++++++++++++++++++++++++++++
 1 file changed, 30 insertions(+)

diff --git a/drivers/iommu/intel/iommu.c b/drivers/iommu/intel/iommu.c
index 1400bf283..cca97e5b2 100644
--- a/drivers/iommu/intel/iommu.c
+++ b/drivers/iommu/intel/iommu.c
@@ -44,6 +44,13 @@
 		((pdev)->vendor == PCI_VENDOR_ID_INTEL && (pdev)->device == 0x34E4) \
 	)
 
+#define IS_INTEL_IPU(pdev) ((pdev)->vendor == PCI_VENDOR_ID_INTEL &&	\
+			   ((pdev)->device == 0x9a19 ||		\
+			    (pdev)->device == 0x9a39 ||		\
+			    (pdev)->device == 0x4e19 ||		\
+			    (pdev)->device == 0x465d ||		\
+			    (pdev)->device == 0x1919))
+
 #define IOAPIC_RANGE_START	(0xfee00000)
 #define IOAPIC_RANGE_END	(0xfeefffff)
 #define IOVA_START_ADDR		(0x1000)
@@ -208,12 +215,14 @@ int intel_iommu_enabled = 0;
 EXPORT_SYMBOL_GPL(intel_iommu_enabled);
 
 static int dmar_map_ipts = 1;
+static int dmar_map_ipu = 1;
 static int intel_iommu_superpage = 1;
 static int iommu_identity_mapping;
 static int iommu_skip_te_disable;
 static int disable_igfx_iommu;
 
 #define IDENTMAP_AZALIA		4
+#define IDENTMAP_IPU		8
 #define IDENTMAP_IPTS		16
 
 const struct iommu_ops intel_iommu_ops;
@@ -1406,6 +1415,9 @@ static int device_def_domain_type(struct device *dev)
 		if ((iommu_identity_mapping & IDENTMAP_AZALIA) && IS_AZALIA(pdev))
 			return IOMMU_DOMAIN_IDENTITY;
 
+		if ((iommu_identity_mapping & IDENTMAP_IPU) && IS_INTEL_IPU(pdev))
+			return IOMMU_DOMAIN_IDENTITY;
+
 		if ((iommu_identity_mapping & IDENTMAP_IPTS) && IS_IPTS(pdev))
 			return IOMMU_DOMAIN_IDENTITY;
 	}
@@ -1698,6 +1710,9 @@ static int __init init_dmars(void)
 		iommu_set_root_entry(iommu);
 	}
 
+	if (!dmar_map_ipu)
+		iommu_identity_mapping |= IDENTMAP_IPU;
+
 	if (!dmar_map_ipts)
 		iommu_identity_mapping |= IDENTMAP_IPTS;
 
@@ -3946,6 +3961,18 @@ static void quirk_iommu_igfx(struct pci_dev *dev)
 	disable_igfx_iommu = 1;
 }
 
+static void quirk_iommu_ipu(struct pci_dev *dev)
+{
+	if (!IS_INTEL_IPU(dev))
+		return;
+
+	if (risky_device(dev))
+		return;
+
+	pci_info(dev, "Passthrough IOMMU for integrated Intel IPU\n");
+	dmar_map_ipu = 0;
+}
+
 static void quirk_iommu_ipts(struct pci_dev *dev)
 {
 	if (!IS_IPTS(dev))
@@ -3996,6 +4023,9 @@ DECLARE_PCI_FIXUP_HEADER(PCI_VENDOR_ID_INTEL, 0x1632, quirk_iommu_igfx);
 DECLARE_PCI_FIXUP_HEADER(PCI_VENDOR_ID_INTEL, 0x163A, quirk_iommu_igfx);
 DECLARE_PCI_FIXUP_HEADER(PCI_VENDOR_ID_INTEL, 0x163D, quirk_iommu_igfx);
 
+/* disable IPU dmar support */
+DECLARE_PCI_FIXUP_HEADER(PCI_VENDOR_ID_INTEL, PCI_ANY_ID, quirk_iommu_ipu);
+
 /* disable IPTS dmar support */
 DECLARE_PCI_FIXUP_HEADER(PCI_VENDOR_ID_INTEL, 0x9D3E, quirk_iommu_ipts);
 DECLARE_PCI_FIXUP_HEADER(PCI_VENDOR_ID_INTEL, 0x34E4, quirk_iommu_ipts);
-- 
2.53.0

From e161e5012e6858f2715f72668ea2fc0d7aeb89da Mon Sep 17 00:00:00 2001
From: Daniel Scally <djrscally@gmail.com>
Date: Sun, 10 Oct 2021 20:57:02 +0200
Subject: [PATCH] platform/x86: int3472: Enable I2c daisy chain

The TPS68470 PMIC has an I2C passthrough mode through which I2C traffic
can be forwarded to a device connected to the PMIC as though it were
connected directly to the system bus. Enable this mode when the chip
is initialised.

Signed-off-by: Daniel Scally <djrscally@gmail.com>
Patchset: cameras
---
 drivers/platform/x86/intel/int3472/tps68470.c | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/drivers/platform/x86/intel/int3472/tps68470.c b/drivers/platform/x86/intel/int3472/tps68470.c
index 013340569..9e0763bdc 100644
--- a/drivers/platform/x86/intel/int3472/tps68470.c
+++ b/drivers/platform/x86/intel/int3472/tps68470.c
@@ -46,6 +46,13 @@ static int tps68470_chip_init(struct device *dev, struct regmap *regmap)
 		return ret;
 	}
 
+	/* Enable I2C daisy chain */
+	ret = regmap_write(regmap, TPS68470_REG_S_I2C_CTL, 0x03);
+	if (ret) {
+		dev_err(dev, "Failed to enable i2c daisy chain\n");
+		return ret;
+	}
+
 	dev_info(dev, "TPS68470 REVID: 0x%02x\n", version);
 
 	return 0;
-- 
2.53.0

From 9c649457de8b93596baaa80f4f8c00e3d9baba20 Mon Sep 17 00:00:00 2001
From: Daniel Scally <dan.scally@ideasonboard.com>
Date: Tue, 21 Mar 2023 13:45:26 +0000
Subject: [PATCH] media: i2c: Clarify that gain is Analogue gain in OV7251

Update the control ID for the gain control in the ov7251 driver to
V4L2_CID_ANALOGUE_GAIN.

Signed-off-by: Daniel Scally <dan.scally@ideasonboard.com>
Patchset: cameras
---
 drivers/media/i2c/ov7251.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/drivers/media/i2c/ov7251.c b/drivers/media/i2c/ov7251.c
index 27afc3fc0..28b192c9a 100644
--- a/drivers/media/i2c/ov7251.c
+++ b/drivers/media/i2c/ov7251.c
@@ -1053,7 +1053,7 @@ static int ov7251_s_ctrl(struct v4l2_ctrl *ctrl)
 	case V4L2_CID_EXPOSURE:
 		ret = ov7251_set_exposure(ov7251, ctrl->val);
 		break;
-	case V4L2_CID_GAIN:
+	case V4L2_CID_ANALOGUE_GAIN:
 		ret = ov7251_set_gain(ov7251, ctrl->val);
 		break;
 	case V4L2_CID_TEST_PATTERN:
@@ -1579,7 +1579,7 @@ static int ov7251_init_ctrls(struct ov7251 *ov7251)
 	ov7251->exposure = v4l2_ctrl_new_std(&ov7251->ctrls, &ov7251_ctrl_ops,
 					     V4L2_CID_EXPOSURE, 1, 32, 1, 32);
 	ov7251->gain = v4l2_ctrl_new_std(&ov7251->ctrls, &ov7251_ctrl_ops,
-					 V4L2_CID_GAIN, 16, 1023, 1, 16);
+					 V4L2_CID_ANALOGUE_GAIN, 16, 1023, 1, 16);
 	v4l2_ctrl_new_std_menu_items(&ov7251->ctrls, &ov7251_ctrl_ops,
 				     V4L2_CID_TEST_PATTERN,
 				     ARRAY_SIZE(ov7251_test_pattern_menu) - 1,
-- 
2.53.0

From a83903189bc92cdd72dd802072f2518bbad79ee3 Mon Sep 17 00:00:00 2001
From: Daniel Scally <dan.scally@ideasonboard.com>
Date: Wed, 22 Mar 2023 11:01:42 +0000
Subject: [PATCH] media: v4l2-core: Acquire privacy led in
 v4l2_async_register_subdev()

The current call to v4l2_subdev_get_privacy_led() is contained in
v4l2_async_register_subdev_sensor(), but that function isn't used by
all the sensor drivers. Move the acquisition of the privacy led to
v4l2_async_register_subdev() instead.

Signed-off-by: Daniel Scally <dan.scally@ideasonboard.com>
Patchset: cameras
---
 drivers/media/v4l2-core/v4l2-async.c  | 4 ++++
 drivers/media/v4l2-core/v4l2-fwnode.c | 4 ----
 2 files changed, 4 insertions(+), 4 deletions(-)

diff --git a/drivers/media/v4l2-core/v4l2-async.c b/drivers/media/v4l2-core/v4l2-async.c
index 1c08bba9e..57a990d7d 100644
--- a/drivers/media/v4l2-core/v4l2-async.c
+++ b/drivers/media/v4l2-core/v4l2-async.c
@@ -811,6 +811,10 @@ int __v4l2_async_register_subdev(struct v4l2_subdev *sd, struct module *module)
 
 	INIT_LIST_HEAD(&sd->asc_list);
 
+	ret = v4l2_subdev_get_privacy_led(sd);
+	if (ret < 0)
+		return ret;
+
 	/*
 	 * No reference taken. The reference is held by the device (struct
 	 * v4l2_subdev.dev), and async sub-device does not exist independently
diff --git a/drivers/media/v4l2-core/v4l2-fwnode.c b/drivers/media/v4l2-core/v4l2-fwnode.c
index cb153ce42..f11b499e1 100644
--- a/drivers/media/v4l2-core/v4l2-fwnode.c
+++ b/drivers/media/v4l2-core/v4l2-fwnode.c
@@ -1260,10 +1260,6 @@ int v4l2_async_register_subdev_sensor(struct v4l2_subdev *sd)
 
 	v4l2_async_subdev_nf_init(notifier, sd);
 
-	ret = v4l2_subdev_get_privacy_led(sd);
-	if (ret < 0)
-		goto out_cleanup;
-
 	ret = v4l2_async_nf_parse_fwnode_sensor(sd->dev, notifier);
 	if (ret < 0)
 		goto out_cleanup;
-- 
2.53.0

From bd907090c2084997d86042ba1961a066e5f6cce8 Mon Sep 17 00:00:00 2001
From: Kate Hsuan <hpa@redhat.com>
Date: Tue, 21 Mar 2023 23:37:16 +0800
Subject: [PATCH] platform: x86: int3472: Add MFD cell for tps68470 LED

Add MFD cell for tps68470-led.

Reviewed-by: Daniel Scally <dan.scally@ideasonboard.com>
Signed-off-by: Kate Hsuan <hpa@redhat.com>
Reviewed-by: Hans de Goede <hdegoede@redhat.com>
Patchset: cameras
---
 drivers/platform/x86/intel/int3472/tps68470.c | 5 +++--
 1 file changed, 3 insertions(+), 2 deletions(-)

diff --git a/drivers/platform/x86/intel/int3472/tps68470.c b/drivers/platform/x86/intel/int3472/tps68470.c
index 9e0763bdc..0976b2679 100644
--- a/drivers/platform/x86/intel/int3472/tps68470.c
+++ b/drivers/platform/x86/intel/int3472/tps68470.c
@@ -17,7 +17,7 @@
 #define DESIGNED_FOR_CHROMEOS		1
 #define DESIGNED_FOR_WINDOWS		2
 
-#define TPS68470_WIN_MFD_CELL_COUNT	3
+#define TPS68470_WIN_MFD_CELL_COUNT	4
 
 static const struct mfd_cell tps68470_cros[] = {
 	{ .name = "tps68470-gpio" },
@@ -203,7 +203,8 @@ static int skl_int3472_tps68470_probe(struct i2c_client *client)
 		cells[1].name = "tps68470-regulator";
 		cells[1].platform_data = (void *)board_data->tps68470_regulator_pdata;
 		cells[1].pdata_size = sizeof(struct tps68470_regulator_platform_data);
-		cells[2].name = "tps68470-gpio";
+		cells[2].name = "tps68470-led";
+		cells[3].name = "tps68470-gpio";
 
 		for (i = 0; i < board_data->n_gpiod_lookups; i++)
 			gpiod_add_lookup_table(board_data->tps68470_gpio_lookup_tables[i]);
-- 
2.53.0

From b4551d25d8d3b88500baa0b2d0cd3e5b60438daf Mon Sep 17 00:00:00 2001
From: Kate Hsuan <hpa@redhat.com>
Date: Tue, 21 Mar 2023 23:37:17 +0800
Subject: [PATCH] include: mfd: tps68470: Add masks for LEDA and LEDB

Add flags for both LEDA(TPS68470_ILEDCTL_ENA), LEDB
(TPS68470_ILEDCTL_ENB), and current control mask for LEDB
(TPS68470_ILEDCTL_CTRLB)

Reviewed-by: Daniel Scally <dan.scally@ideasonboard.com>
Reviewed-by: Hans de Goede <hdegoede@redhat.com>
Signed-off-by: Kate Hsuan <hpa@redhat.com>
Patchset: cameras
---
 include/linux/mfd/tps68470.h | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/include/linux/mfd/tps68470.h b/include/linux/mfd/tps68470.h
index 7807fa329..2d2abb25b 100644
--- a/include/linux/mfd/tps68470.h
+++ b/include/linux/mfd/tps68470.h
@@ -34,6 +34,7 @@
 #define TPS68470_REG_SGPO		0x22
 #define TPS68470_REG_GPDI		0x26
 #define TPS68470_REG_GPDO		0x27
+#define TPS68470_REG_ILEDCTL		0x28
 #define TPS68470_REG_VCMVAL		0x3C
 #define TPS68470_REG_VAUX1VAL		0x3D
 #define TPS68470_REG_VAUX2VAL		0x3E
@@ -94,4 +95,8 @@
 #define TPS68470_GPIO_MODE_OUT_CMOS	2
 #define TPS68470_GPIO_MODE_OUT_ODRAIN	3
 
+#define TPS68470_ILEDCTL_ENA		BIT(2)
+#define TPS68470_ILEDCTL_ENB		BIT(6)
+#define TPS68470_ILEDCTL_CTRLB		GENMASK(5, 4)
+
 #endif /* __LINUX_MFD_TPS68470_H */
-- 
2.53.0

From 3a215fa6681957570a03d2fa2a936c4400a0248c Mon Sep 17 00:00:00 2001
From: Kate Hsuan <hpa@redhat.com>
Date: Tue, 21 Mar 2023 23:37:18 +0800
Subject: [PATCH] leds: tps68470: Add LED control for tps68470

There are two LED controllers, LEDA indicator LED and LEDB flash LED for
tps68470. LEDA can be enabled by setting TPS68470_ILEDCTL_ENA. Moreover,
tps68470 provides four levels of power status for LEDB. If the
properties called "ti,ledb-current" can be found, the current will be
set according to the property values. These two LEDs can be controlled
through the LED class of sysfs (tps68470-leda and tps68470-ledb).

Signed-off-by: Kate Hsuan <hpa@redhat.com>
Reviewed-by: Hans de Goede <hdegoede@redhat.com>
Patchset: cameras
---
 drivers/leds/Kconfig         |  12 +++
 drivers/leds/Makefile        |   1 +
 drivers/leds/leds-tps68470.c | 185 +++++++++++++++++++++++++++++++++++
 3 files changed, 198 insertions(+)
 create mode 100644 drivers/leds/leds-tps68470.c

diff --git a/drivers/leds/Kconfig b/drivers/leds/Kconfig
index 11e7282dc..c5665bd5e 100644
--- a/drivers/leds/Kconfig
+++ b/drivers/leds/Kconfig
@@ -998,6 +998,18 @@ config LEDS_TPS6105X
 	  It is a single boost converter primarily for white LEDs and
 	  audio amplifiers.
 
+config LEDS_TPS68470
+	tristate "LED support for TI TPS68470"
+	depends on LEDS_CLASS
+	depends on INTEL_SKL_INT3472
+	help
+	  This driver supports TPS68470 PMIC with LED chip.
+	  It provides two LED controllers, with the ability to drive 2
+	  indicator LEDs and 2 flash LEDs.
+
+	  To compile this driver as a module, choose M and it will be
+	  called leds-tps68470
+
 config LEDS_IP30
 	tristate "LED support for SGI Octane machines"
 	depends on LEDS_CLASS
diff --git a/drivers/leds/Makefile b/drivers/leds/Makefile
index 9a0333ec1..4e1ca8fc9 100644
--- a/drivers/leds/Makefile
+++ b/drivers/leds/Makefile
@@ -93,6 +93,7 @@ obj-$(CONFIG_LEDS_TCA6507)		+= leds-tca6507.o
 obj-$(CONFIG_LEDS_TI_LMU_COMMON)	+= leds-ti-lmu-common.o
 obj-$(CONFIG_LEDS_TLC591XX)		+= leds-tlc591xx.o
 obj-$(CONFIG_LEDS_TPS6105X)		+= leds-tps6105x.o
+obj-$(CONFIG_LEDS_TPS68470)		+= leds-tps68470.o
 obj-$(CONFIG_LEDS_TURRIS_OMNIA)		+= leds-turris-omnia.o
 obj-$(CONFIG_LEDS_UPBOARD)		+= leds-upboard.o
 obj-$(CONFIG_LEDS_WM831X_STATUS)	+= leds-wm831x-status.o
diff --git a/drivers/leds/leds-tps68470.c b/drivers/leds/leds-tps68470.c
new file mode 100644
index 000000000..35aeb5db8
--- /dev/null
+++ b/drivers/leds/leds-tps68470.c
@@ -0,0 +1,185 @@
+// SPDX-License-Identifier: GPL-2.0
+/*
+ * LED driver for TPS68470 PMIC
+ *
+ * Copyright (C) 2023 Red Hat
+ *
+ * Authors:
+ *	Kate Hsuan <hpa@redhat.com>
+ */
+
+#include <linux/leds.h>
+#include <linux/mfd/tps68470.h>
+#include <linux/module.h>
+#include <linux/platform_device.h>
+#include <linux/property.h>
+#include <linux/regmap.h>
+
+
+#define lcdev_to_led(led_cdev) \
+	container_of(led_cdev, struct tps68470_led, lcdev)
+
+#define led_to_tps68470(led, index) \
+	container_of(led, struct tps68470_device, leds[index])
+
+enum tps68470_led_ids {
+	TPS68470_ILED_A,
+	TPS68470_ILED_B,
+	TPS68470_NUM_LEDS
+};
+
+static const char *tps68470_led_names[] = {
+	[TPS68470_ILED_A] = "tps68470-iled_a",
+	[TPS68470_ILED_B] = "tps68470-iled_b",
+};
+
+struct tps68470_led {
+	unsigned int led_id;
+	struct led_classdev lcdev;
+};
+
+struct tps68470_device {
+	struct device *dev;
+	struct regmap *regmap;
+	struct tps68470_led leds[TPS68470_NUM_LEDS];
+};
+
+enum ctrlb_current {
+	CTRLB_2MA	= 0,
+	CTRLB_4MA	= 1,
+	CTRLB_8MA	= 2,
+	CTRLB_16MA	= 3,
+};
+
+static int tps68470_brightness_set(struct led_classdev *led_cdev, enum led_brightness brightness)
+{
+	struct tps68470_led *led = lcdev_to_led(led_cdev);
+	struct tps68470_device *tps68470 = led_to_tps68470(led, led->led_id);
+	struct regmap *regmap = tps68470->regmap;
+
+	switch (led->led_id) {
+	case TPS68470_ILED_A:
+		return regmap_update_bits(regmap, TPS68470_REG_ILEDCTL, TPS68470_ILEDCTL_ENA,
+					  brightness ? TPS68470_ILEDCTL_ENA : 0);
+	case TPS68470_ILED_B:
+		return regmap_update_bits(regmap, TPS68470_REG_ILEDCTL, TPS68470_ILEDCTL_ENB,
+					  brightness ? TPS68470_ILEDCTL_ENB : 0);
+	}
+	return -EINVAL;
+}
+
+static enum led_brightness tps68470_brightness_get(struct led_classdev *led_cdev)
+{
+	struct tps68470_led *led = lcdev_to_led(led_cdev);
+	struct tps68470_device *tps68470 = led_to_tps68470(led, led->led_id);
+	struct regmap *regmap = tps68470->regmap;
+	int ret = 0;
+	int value = 0;
+
+	ret =  regmap_read(regmap, TPS68470_REG_ILEDCTL, &value);
+	if (ret)
+		return dev_err_probe(led_cdev->dev, -EINVAL, "failed on reading register\n");
+
+	switch (led->led_id) {
+	case TPS68470_ILED_A:
+		value = value & TPS68470_ILEDCTL_ENA;
+		break;
+	case TPS68470_ILED_B:
+		value = value & TPS68470_ILEDCTL_ENB;
+		break;
+	}
+
+	return value ? LED_ON : LED_OFF;
+}
+
+
+static int tps68470_ledb_current_init(struct platform_device *pdev,
+				      struct tps68470_device *tps68470)
+{
+	int ret = 0;
+	unsigned int curr;
+
+	/* configure LEDB current if the properties can be got */
+	if (!device_property_read_u32(&pdev->dev, "ti,ledb-current", &curr)) {
+		if (curr > CTRLB_16MA) {
+			dev_err(&pdev->dev,
+				"Invalid LEDB current value: %d\n",
+				curr);
+			return -EINVAL;
+		}
+		ret = regmap_update_bits(tps68470->regmap, TPS68470_REG_ILEDCTL,
+					 TPS68470_ILEDCTL_CTRLB, curr);
+	}
+	return ret;
+}
+
+static int tps68470_leds_probe(struct platform_device *pdev)
+{
+	int i = 0;
+	int ret = 0;
+	struct tps68470_device *tps68470;
+	struct tps68470_led *led;
+	struct led_classdev *lcdev;
+
+	tps68470 = devm_kzalloc(&pdev->dev, sizeof(struct tps68470_device),
+				GFP_KERNEL);
+	if (!tps68470)
+		return -ENOMEM;
+
+	tps68470->dev = &pdev->dev;
+	tps68470->regmap = dev_get_drvdata(pdev->dev.parent);
+
+	for (i = 0; i < TPS68470_NUM_LEDS; i++) {
+		led = &tps68470->leds[i];
+		lcdev = &led->lcdev;
+
+		led->led_id = i;
+
+		lcdev->name = devm_kasprintf(tps68470->dev, GFP_KERNEL, "%s::%s",
+					     tps68470_led_names[i], LED_FUNCTION_INDICATOR);
+		if (!lcdev->name)
+			return -ENOMEM;
+
+		lcdev->max_brightness = 1;
+		lcdev->brightness = 0;
+		lcdev->brightness_set_blocking = tps68470_brightness_set;
+		lcdev->brightness_get = tps68470_brightness_get;
+		lcdev->dev = &pdev->dev;
+
+		ret = devm_led_classdev_register(tps68470->dev, lcdev);
+		if (ret) {
+			dev_err_probe(tps68470->dev, ret,
+				      "error registering led\n");
+			goto err_exit;
+		}
+
+		if (i == TPS68470_ILED_B) {
+			ret = tps68470_ledb_current_init(pdev, tps68470);
+			if (ret)
+				goto err_exit;
+		}
+	}
+
+err_exit:
+	if (ret) {
+		for (i = 0; i < TPS68470_NUM_LEDS; i++) {
+			if (tps68470->leds[i].lcdev.name)
+				devm_led_classdev_unregister(&pdev->dev,
+							     &tps68470->leds[i].lcdev);
+		}
+	}
+
+	return ret;
+}
+static struct platform_driver tps68470_led_driver = {
+	.driver = {
+		   .name = "tps68470-led",
+	},
+	.probe = tps68470_leds_probe,
+};
+
+module_platform_driver(tps68470_led_driver);
+
+MODULE_ALIAS("platform:tps68470-led");
+MODULE_DESCRIPTION("LED driver for TPS68470 PMIC");
+MODULE_LICENSE("GPL v2");
-- 
2.53.0

From b581a619711c71600e9f083e3f8d3a36092cb11f Mon Sep 17 00:00:00 2001
From: mojyack <mojyack@gmail.com>
Date: Tue, 26 Mar 2024 05:55:44 +0900
Subject: [PATCH] media: i2c: dw9719: fix probe error on surface go 2

On surface go 2, sometimes probing dw9719 fails with "dw9719: probe of i2c-INT347A:00-VCM failed with error -121".
The -121(-EREMOTEIO) is came from drivers/i2c/busses/i2c-designware-common.c:575, and indicates the initialize occurs too early.
So just add some delay.
There is no exact reason for this 10000us, but 100us failed.

Patchset: cameras
---
 drivers/media/i2c/dw9719.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/drivers/media/i2c/dw9719.c b/drivers/media/i2c/dw9719.c
index 595583359..b89765521 100644
--- a/drivers/media/i2c/dw9719.c
+++ b/drivers/media/i2c/dw9719.c
@@ -121,6 +121,9 @@ static int dw9719_power_up(struct dw9719_device *dw9719, bool detect)
 	if (ret)
 		return ret;
 
+	/* Wait for device to be acknowledged */
+	fsleep(10000);
+
 	/*
 	 * Need 100us to transition from SHUTDOWN to STANDBY.
 	 * Jiggle the SCL pin to wake up the device (even when the regulator is
-- 
2.53.0

From b1dc230fffb35a88809f6c7877416887c1fb3d20 Mon Sep 17 00:00:00 2001
From: Tooraj Taraz <tooraj.taraz@yahoo.com>
Date: Wed, 31 Dec 2025 00:35:50 +0100
Subject: [PATCH] Add camera support for Surface Pro 9

Experimental camera support for the Surface Pro 9.

Link: https://github.com/linux-surface/linux-surface/pull/1867
Patchset: cameras
---
 drivers/media/i2c/ov13858.c                   | 158 +++++++++++++++++-
 drivers/media/i2c/ov5693.c                    |   1 +
 drivers/media/pci/intel/ipu-bridge.c          |   4 +
 drivers/platform/x86/intel/int3472/discrete.c |  22 +++
 include/linux/platform_data/x86/int3472.h     |   4 ++--
 5 files changed, 185 insertions(+), 4 deletions(-)

diff --git a/drivers/media/i2c/ov13858.c b/drivers/media/i2c/ov13858.c
index 162b49046..85cd0c98a 100644
--- a/drivers/media/i2c/ov13858.c
+++ b/drivers/media/i2c/ov13858.c
@@ -2,6 +2,7 @@
 // Copyright (c) 2017 Intel Corporation.
 
 #include <linux/acpi.h>
+#include <linux/delay.h>
 #include <linux/clk.h>
 #include <linux/i2c.h>
 #include <linux/module.h>
@@ -119,6 +120,14 @@ struct ov13858_mode {
 	struct ov13858_reg_list reg_list;
 };
 
+// Or use standard names that might be more correct:
+static const char * const ov13858_supply_names[] = {
+	"avdd",    // Analog voltage
+	"pwr1",    // Digital core voltage
+};
+
+#define OV13858_NUM_SUPPLIES ARRAY_SIZE(ov13858_supply_names)
+
 /* 4224x3136 needs 1080Mbps/lane, 4 lanes */
 static const struct ov13858_reg mipi_data_rate_1080mbps[] = {
 	/* PLL1 registers */
@@ -1046,12 +1055,126 @@ struct ov13858 {
 	/* Current mode */
 	const struct ov13858_mode *cur_mode;
 
+	struct regulator_bulk_data supplies[OV13858_NUM_SUPPLIES];
+	struct gpio_desc *reset;
+	struct clk *xvclk;
+
 	/* Mutex for serialized access */
 	struct mutex mutex;
 };
 
 #define to_ov13858(_sd)	container_of(_sd, struct ov13858, sd)
 
+
+static int ov13858_get_regulators(struct ov13858 *ov13858)
+{
+	unsigned int i;
+
+	for (i = 0; i < OV13858_NUM_SUPPLIES; i++)
+		ov13858->supplies[i].supply = ov13858_supply_names[i];
+
+	return devm_regulator_bulk_get(ov13858->dev, OV13858_NUM_SUPPLIES,
+				       ov13858->supplies);
+}
+
+static int ov13858_sensor_powerup(struct ov13858 *ov13858)
+{
+	int ret;
+
+	dev_info(ov13858->dev, "Powering up sensor\n");
+
+	/* Assert reset */
+	gpiod_set_value_cansleep(ov13858->reset, 1);
+
+	/* Enable clock FIRST - sensor needs clock to communicate */
+	ret = clk_prepare_enable(ov13858->xvclk);
+	if (ret) {
+		dev_err(ov13858->dev, "Failed to enable clock: %d\n", ret);
+		return ret;
+	}
+	dev_info(ov13858->dev, "Clock enabled\n");
+
+	/* Enable regulators */
+	ret = regulator_bulk_enable(OV13858_NUM_SUPPLIES, ov13858->supplies);
+	if (ret) {
+		dev_err(ov13858->dev, "Failed to enable regulators: %d\n", ret);
+		clk_disable_unprepare(ov13858->xvclk);
+		return ret;
+	}
+
+	dev_info(ov13858->dev, "Regulators enabled\n");
+
+	/* Wait for power to stabilize */
+	usleep_range(5000, 10000);
+
+	/* De-assert reset */
+	gpiod_set_value_cansleep(ov13858->reset, 0);
+
+	/* Wait for sensor to boot */
+	msleep(20);
+
+	dev_info(ov13858->dev, "Reset de-asserted, sensor should be ready\n");
+
+	return 0;
+}
+
+static void ov13858_sensor_powerdown(struct ov13858 *ov13858)
+{
+	dev_info(ov13858->dev, "Powering down sensor\n");
+
+	/* Assert reset to put sensor in reset state */
+	gpiod_set_value_cansleep(ov13858->reset, 1);
+
+	/* Disable regulators */
+	regulator_bulk_disable(OV13858_NUM_SUPPLIES, ov13858->supplies);
+	dev_info(ov13858->dev, "Regulators disabled\n");
+
+	/* Disable clock */
+	clk_disable_unprepare(ov13858->xvclk);
+	dev_info(ov13858->dev, "Clock disabled\n");
+}
+
+static int __maybe_unused ov13858_suspend(struct device *dev)
+{
+	struct i2c_client *client = to_i2c_client(dev);
+	struct v4l2_subdev *sd = i2c_get_clientdata(client);
+	struct ov13858 *ov13858 = to_ov13858(sd);
+
+	ov13858_sensor_powerdown(ov13858);
+
+	return 0;
+}
+
+static int __maybe_unused ov13858_resume(struct device *dev)
+{
+	struct i2c_client *client = to_i2c_client(dev);
+	struct v4l2_subdev *sd = i2c_get_clientdata(client);
+	struct ov13858 *ov13858 = to_ov13858(sd);
+	int ret;
+
+	ret = ov13858_sensor_powerup(ov13858);
+	if (ret)
+		return ret;
+
+	return 0;
+}
+
+static const struct dev_pm_ops ov13858_pm_ops = {
+	SET_RUNTIME_PM_OPS(ov13858_suspend, ov13858_resume, NULL)
+};
+
+static int ov13858_get_gpios(struct ov13858 *ov13858)
+{
+	ov13858->reset = devm_gpiod_get_optional(ov13858->dev, "reset",
+	                                         GPIOD_OUT_HIGH);
+	if (IS_ERR(ov13858->reset)) {
+		dev_err(ov13858->dev, "Error fetching reset GPIO\n");
+		return PTR_ERR(ov13858->reset);
+	}
+
+	return 0;
+}
+
 /* Read registers up to 4 at a time */
 static int ov13858_read_reg(struct ov13858 *ov13858, u16 reg, u32 len,
 			    u32 *val)
@@ -1498,8 +1621,11 @@ static int ov13858_identify_module(struct ov13858 *ov13858)
 	int ret;
 	u32 val;
 
+	dev_info(ov13858->dev, "Attempting to read chip ID from register 0x300a\n");
 	ret = ov13858_read_reg(ov13858, OV13858_REG_CHIP_ID,
 			       OV13858_REG_VALUE_24BIT, &val);
+	dev_info(ov13858->dev, "Chip ID read result: ret=%d, val=0x%06x (expected 0x%06x)\n",
+	         ret, val, OV13858_CHIP_ID);
 	if (ret)
 		return ret;
 
@@ -1509,6 +1635,7 @@ static int ov13858_identify_module(struct ov13858 *ov13858)
 		return -EIO;
 	}
 
+	dev_info(ov13858->dev, "Chip ID verified successfully!\n");
 	return 0;
 }
 
@@ -1678,14 +1805,33 @@ static int ov13858_probe(struct i2c_client *client)
 				     "external clock %lu is not supported\n",
 				     freq);
 
+	/* Get the external clock */
+	ov13858->xvclk = devm_clk_get_optional(ov13858->dev, "xvclk");
+	if (IS_ERR(ov13858->xvclk))
+		return dev_err_probe(ov13858->dev, PTR_ERR(ov13858->xvclk),
+		       "failed to get xvclk\n");
+
 	/* Initialize subdev */
 	v4l2_i2c_subdev_init(&ov13858->sd, client, &ov13858_subdev_ops);
 
+	ret = ov13858_get_regulators(ov13858);
+	if (ret)
+		return dev_err_probe(ov13858->dev, ret,
+		       "Error fetching regulators\n");
+
+	ret = ov13858_get_gpios(ov13858);
+	if (ret)
+		return ret;
+
+	ret = ov13858_sensor_powerup(ov13858);
+	if (ret)
+		return ret;  // No cleanup needed yet, devm handles GPIOs/regulators
+
 	/* Check module identity */
 	ret = ov13858_identify_module(ov13858);
 	if (ret) {
 		dev_err(ov13858->dev, "failed to find sensor: %d\n", ret);
-		return ret;
+		goto error_power_off;  // Now we need to power down
 	}
 
 	/* Set default mode to max resolution */
@@ -1693,7 +1839,7 @@ static int ov13858_probe(struct i2c_client *client)
 
 	ret = ov13858_init_controls(ov13858);
 	if (ret)
-		return ret;
+		goto error_power_off;
 
 	/* Initialize subdev */
 	ov13858->sd.internal_ops = &ov13858_internal_ops;
@@ -1729,6 +1875,10 @@ static int ov13858_probe(struct i2c_client *client)
 
 error_handler_free:
 	ov13858_free_controls(ov13858);
+
+error_power_off:
+	ov13858_sensor_powerdown(ov13858);
+
 	dev_err(ov13858->dev, "%s failed:%d\n", __func__, ret);
 
 	return ret;
@@ -1744,6 +1894,9 @@ static void ov13858_remove(struct i2c_client *client)
 	ov13858_free_controls(ov13858);
 
 	pm_runtime_disable(ov13858->dev);
+	if (!pm_runtime_status_suspended(ov13858->dev))
+		ov13858_sensor_powerdown(ov13858);
+	pm_runtime_set_suspended(ov13858->dev);
 }
 
 static const struct i2c_device_id ov13858_id_table[] = {
@@ -1765,6 +1918,7 @@ MODULE_DEVICE_TABLE(acpi, ov13858_acpi_ids);
 static struct i2c_driver ov13858_i2c_driver = {
 	.driver = {
 		.name = "ov13858",
+		.pm = &ov13858_pm_ops,
 		.acpi_match_table = ACPI_PTR(ov13858_acpi_ids),
 	},
 	.probe = ov13858_probe,
diff --git a/drivers/media/i2c/ov5693.c b/drivers/media/i2c/ov5693.c
index 4cc796bbe..02236f3db 100644
--- a/drivers/media/i2c/ov5693.c
+++ b/drivers/media/i2c/ov5693.c
@@ -1396,6 +1396,7 @@ static const struct dev_pm_ops ov5693_pm_ops = {
 
 static const struct acpi_device_id ov5693_acpi_match[] = {
 	{"INT33BE"},
+	{"OVTI5693"},
 	{},
 };
 MODULE_DEVICE_TABLE(acpi, ov5693_acpi_match);
diff --git a/drivers/media/pci/intel/ipu-bridge.c b/drivers/media/pci/intel/ipu-bridge.c
index b2b710094..f5456330b 100644
--- a/drivers/media/pci/intel/ipu-bridge.c
+++ b/drivers/media/pci/intel/ipu-bridge.c
@@ -61,6 +61,10 @@ static const struct ipu_sensor_config ipu_supported_sensors[] = {
 	IPU_SENSOR_CONFIG("INT33BE", 1, 419200000),
 	/* Onsemi MT9M114 */
 	IPU_SENSOR_CONFIG("INT33F0", 1, 384000000),
+	/* Omnivision OV5693 - Surface Pro 9 */
+	IPU_SENSOR_CONFIG("OVTI5693", 1, 419200000),
+	/* Omnivision OV13858 - Surface Pro 9 */
+	IPU_SENSOR_CONFIG("OVTID858", 4, 540000000),
 	/* Omnivision OV2740 */
 	IPU_SENSOR_CONFIG("INT3474", 1, 180000000),
 	/* Omnivision OV5670 */
diff --git a/drivers/platform/x86/intel/int3472/discrete.c b/drivers/platform/x86/intel/int3472/discrete.c
index 1505fc3ef..f0b72b1c1 100644
--- a/drivers/platform/x86/intel/int3472/discrete.c
+++ b/drivers/platform/x86/intel/int3472/discrete.c
@@ -229,6 +229,14 @@ static void int3472_get_con_id_and_polarity(struct int3472_discrete_device *int3
 		/* Setups using a handshake pin need 25 ms enable delay */
 		*enable_time_us = 25 * USEC_PER_MSEC;
 		break;
+	case 0x08:  /* Surface Pro 9 - additional power rail */
+		*con_id = "pwr1";
+		*gpio_flags = GPIO_ACTIVE_HIGH;
+		break;
+	case 0x10:  /* Matches mainline v7.0 INT3472_GPIO_TYPE_DOVDD naming */
+		*con_id = "dovdd";
+		*gpio_flags = GPIO_ACTIVE_HIGH;
+		break;
 	default:
 		*con_id = "unknown";
 		*gpio_flags = GPIO_ACTIVE_HIGH;
@@ -333,6 +341,8 @@ static int skl_int3472_handle_gpio_resources(struct acpi_resource *ares,
 	case INT3472_GPIO_TYPE_PRIVACY_LED:
 	case INT3472_GPIO_TYPE_POWER_ENABLE:
 	case INT3472_GPIO_TYPE_HANDSHAKE:
+	case 0x08:  /* Surface Pro 9 power rails */
+	case 0x10:
 		gpio = skl_int3472_gpiod_get_from_temp_lookup(int3472, agpio, con_id, gpio_flags);
 		if (IS_ERR(gpio)) {
 			ret = PTR_ERR(gpio);
@@ -363,6 +373,18 @@ static int skl_int3472_handle_gpio_resources(struct acpi_resource *ares,
 				err_msg = "Failed to register regulator\n";
 
 			break;
+		case 0x08:  /* Surface Pro 9 - treat as power*/
+		case 0x10:  /* "dovdd" - register a regulator like 0x08 */
+		    dev_info(int3472->dev, "GPIO type 0x%02x detected on pin 0x%02x\n", type, agpio->pin_table[0]);
+		    dev_info(int3472->dev, "  con_id=%s, flags=0x%x\n", con_id, gpio_flags);
+		    ret = skl_int3472_register_regulator(int3472, gpio,
+			 GPIO_REGULATOR_ENABLE_TIME,
+			 con_id, NULL);
+		    dev_info(int3472->dev, "  register_regulator returned: %d\n", ret);
+		    if (ret) {
+			dev_err(int3472->dev, "Failed to register type 0x02x: %d\n", type, ret);
+		    }
+		    break;
 		default: /* Never reached */
 			ret = -EINVAL;
 			break;
diff --git a/include/linux/platform_data/x86/int3472.h b/include/linux/platform_data/x86/int3472.h
index b1b837583..7000dd8f4 100644
--- a/include/linux/platform_data/x86/int3472.h
+++ b/include/linux/platform_data/x86/int3472.h
@@ -33,8 +33,8 @@
 #define INT3472_MAX_SENSOR_GPIOS				3
 #define INT3472_MAX_REGULATORS					3
 
-/* E.g. "avdd\0" */
-#define GPIO_SUPPLY_NAME_LENGTH				5
+/* E.g. "dovdd\0" */
+#define GPIO_SUPPLY_NAME_LENGTH				6
 /* 12 chars for acpi_dev_name() + "-", e.g. "ABCD1234:00-" */
 #define GPIO_REGULATOR_NAME_LENGTH				(12 + GPIO_SUPPLY_NAME_LENGTH)
 /* lower- and upper-case mapping */
-- 
2.53.0

From f51a44f59234d641cd068c4c1ba7c643dfe4c0ef Mon Sep 17 00:00:00 2001
From: gabrielstedman <gabrielstedman@users.noreply.github.com>
Date: Fri, 13 Mar 2026 08:52:07 +0000
Subject: [PATCH] media: ipu-bridge: Add front camera rotation quirks for
 Surface Pro 8 and 9

On the Surface Pro 8 and 9, camera rotation in `ssdb.degree` is `0` (no
rotation) but should be `180`. Add these devices to the respective quirk
list.

Link: https://github.com/linux-surface/kernel/pull/160
Patchset: cameras
---
 drivers/media/pci/intel/ipu-bridge.c | 14 ++++++++++++++
 1 file changed, 14 insertions(+)

diff --git a/drivers/media/pci/intel/ipu-bridge.c b/drivers/media/pci/intel/ipu-bridge.c
index f5456330b..3418ff475 100644
--- a/drivers/media/pci/intel/ipu-bridge.c
+++ b/drivers/media/pci/intel/ipu-bridge.c
@@ -122,6 +122,20 @@ static const struct dmi_system_id upside_down_sensor_dmi_ids[] = {
 		},
 		.driver_data = "OVTI02C1",
 	},
+	{
+		.matches = {
+			DMI_MATCH(DMI_SYS_VENDOR, "Microsoft Corporation"),
+			DMI_MATCH(DMI_PRODUCT_NAME, "Surface Pro 8"),
+		},
+		.driver_data = "INT33BE",
+	},
+	{
+		.matches = {
+			DMI_MATCH(DMI_SYS_VENDOR, "Microsoft Corporation"),
+			DMI_MATCH(DMI_PRODUCT_NAME, "Surface Pro 9"),
+		},
+		.driver_data = "OVTI5693",
+	},
 	{} /* Terminating entry */
 };
 
-- 
2.53.0

From d5253e9181f6d392468dfcd454ffdd98fe7a13ee Mon Sep 17 00:00:00 2001
From: "D. Manresa" <d.manresa@electronica-submarina.com>
Date: Thu, 27 Aug 2026 23:03:33 +0200
Subject: [PATCH] media: i2c: ov5693: make the 2x2-binned mode stream on IPU6

The driver's 1296x972 binned mode has never streamed on IPU6 machines
(Surface Pro 7+/8/9, Go 4): the receiver reports "Transfer FIFO
overflow" or "Frame sync error" and no frame ever completes. Because of
that, every IPU6 setup has had to read the sensor at full 2592x1944 and
downscale, throwing away two stops of light and pinning the AGC at its
maximum analogue gain (127 vs ~15) - the familiar "working but unusably
noisy" IPU6 front camera. Fix the mode by porting what the vendor
(Windows) driver ov5693.sys programs differently, extracted from its
mode tables:

- Per-mode MIPI PLL (0x30b3: 0x70 binned / 0x83 full) and six per-mode
  analogue readout registers. Mainline writes only the full-resolution
  values, once, globally; the binned readout then overflows the
  receiver's FIFO.

- Readout geometry with spare pixels: cropping exactly the output size
  starves the sensor's internal ISP and every frame fails CSI-2 framing
  ("Frame sync error"). The vendor reads the whole pixel array, bins it
  to 1312x978, and places the 1296x972 output via the ISP window
  offsets. The parity of the Y offset selects the vertical Bayer phase
  (even = the advertised BGGR, odd = GRBG); it is exposed as a module
  parameter (binned_y_offset, default 2) because libcamera's software
  ISP wants the advertised phase while pipelines built around the
  Windows tuning (Intel's IPU6 HAL/PSYS) expect the vendor's odd
  offset.

- Mode-dependent vertical-flip bits: with binning, setting both FORMAT1
  flip bits corrupts the CSI-2 framing and the ISP bit alone does not
  flip; only the sensor-side bit both streams and flips. Re-apply the
  vflip control when the mode is configured, since the correct bits
  depend on the selected mode.

- Cap the frame rate at 30 fps: the auto-picked 60 fps VTS for the
  binned mode exceeds what the slower binned MIPI PLL can carry (the
  vendor runs this mode at VTS 1984 / 30 fps).

- Program MIPI_CTRL00 (0x4800) = 0x2d before stream-on. The 0x00
  power-on default is fine on IPU3 but leaves the IPU6 D-PHY unable to
  lock: the sensor streams yet no CSI-2 data arrives.

Also document the black-level pedestal next to the BLC init values: the
sensor keeps ~16 LSB (10-bit) at every analogue gain, while OEM .aiqb
calibrations state 64.9 because Windows' ISYS delivers RAW10 aligned to
12 bits.

Verified on a Surface Pro 7+ (IPU6, Tiger Lake): 1296x972 at 30 fps,
analogue gain in normal indoor light drops from 127 (maximum) to ~15,
both through libcamera's software ISP and through Intel's IPU6 HAL on
top of the hardware ISP. Reference setup and measurements:
https://github.com/dmanresa-saes/surface-ipu6-cameras

Signed-off-by: D. Manresa <d.manresa@electronica-submarina.com>
Co-Authored-By: Claude Fable 5 <noreply@anthropic.com>
Patchset: cameras
---
 drivers/media/i2c/ov5693.c | 168 ++++++++++++++++++++++++++++++++++---
 1 file changed, 156 insertions(+), 12 deletions(-)

diff --git a/drivers/media/i2c/ov5693.c b/drivers/media/i2c/ov5693.c
index 02236f3..ee81b4d 100644
--- a/drivers/media/i2c/ov5693.c
+++ b/drivers/media/i2c/ov5693.c
@@ -82,6 +82,17 @@
 #define OV5693_OFFSET_START_X_REG		CCI_REG16(0x3810)
 #define OV5693_OFFSET_START_Y_REG		CCI_REG16(0x3812)
 
+/*
+ * ISP window Y offset used in the binned mode. Parity selects the Bayer
+ * phase of the output: 2 (even) keeps the advertised BGGR phase and is what
+ * libcamera's software ISP expects; 3 (odd, the Windows value) yields the
+ * GRBG phase that the IPU6 PSYS program groups from the Windows tuning
+ * expect. Switch with: modprobe ov5693 binned_y_offset=3
+ */
+static int binned_y_offset = 2;
+module_param(binned_y_offset, int, 0644);
+MODULE_PARM_DESC(binned_y_offset, "binned-mode ISP window Y offset (2=softISP/BGGR, 3=PSYS/GRBG)");
+
 #define OV5693_SUB_INC_X_REG			CCI_REG8(0x3814)
 #define OV5693_SUB_INC_Y_REG			CCI_REG8(0x3815)
 
@@ -110,6 +121,10 @@
 #define OV5693_MIN_CROP_WIDTH			2
 #define OV5693_MIN_CROP_HEIGHT			2
 
+/* MIPI control */
+#define OV5693_MIPI_CTRL00_REG			CCI_REG8(0x4800)
+#define OV5693_MIPI_CTRL00_IPU6			0x2d
+
 /* Test Pattern */
 #define OV5693_TEST_PATTERN_REG			CCI_REG8(0x5e00)
 #define OV5693_TEST_PATTERN_ENABLE		BIT(7)
@@ -285,6 +300,15 @@ static const struct cci_reg_sequence ov5693_global_regs[] = {
 	{CCI_REG8(0x3a04), 0x06},
 	{CCI_REG8(0x3a05), 0x14},
 	{CCI_REG8(0x3e07), 0x20},
+	/*
+	 * BLC: same values the Windows driver writes. The sensor keeps its
+	 * black pedestal at ~16 LSB (10-bit) at every analogue gain (measured
+	 * with dark captures; adding the remaining Windows BLC registers
+	 * 0x4002/0x4005/0x400c/0x400d/0x404e/0x404f does not change that).
+	 * Note the OEM .aiqb black-level calibration says 64.9 because it is
+	 * expressed at the 12-bit alignment Windows' ISYS DMA delivers
+	 * (16.2 <<2); see REPRODUCIR.md section 10.
+	 */
 	{CCI_REG8(0x4000), 0x08},
 	{CCI_REG8(0x4001), 0x04},
 	{CCI_REG8(0x4004), 0x08},
@@ -353,15 +377,36 @@ static const u8 ov5693_test_pattern_bits[] = {
 
 /* V4L2 Controls Functions */
 
+/*
+ * Which FORMAT1 bits express a vertical flip depends on the readout mode.
+ *
+ * At full resolution both the sensor-side and the ISP-side bit are used, as
+ * the datasheet describes. With 2x2 binning, setting both corrupts the CSI-2
+ * framing: an IPU6 receiver reports "Frame sync error" on every frame and no
+ * data arrives at all. The ISP bit alone streams cleanly but does not
+ * actually flip the picture, so the sensor-side bit alone is the one that
+ * both streams and flips. It also inverts the vertical Bayer phase, which
+ * the even ISP window offset in ov5693_mode_configure() keeps in place --
+ * measured on a Surface Pro 7+ against the full-resolution orientation.
+ */
+static u8 ov5693_flip_vert_bits(struct ov5693_device *ov5693)
+{
+	if (ov5693->mode.binning_y)
+		return OV5693_FORMAT1_FLIP_VERT_SENSOR_EN;
+
+	return OV5693_FORMAT1_FLIP_VERT_ISP_EN |
+	       OV5693_FORMAT1_FLIP_VERT_SENSOR_EN;
+}
+
 static int ov5693_flip_vert_configure(struct ov5693_device *ov5693,
 				      bool enable)
 {
-	u8 bits = OV5693_FORMAT1_FLIP_VERT_ISP_EN |
+	u8 mask = OV5693_FORMAT1_FLIP_VERT_ISP_EN |
 		  OV5693_FORMAT1_FLIP_VERT_SENSOR_EN;
 	int ret;
 
-	ret = cci_update_bits(ov5693->regmap, OV5693_FORMAT1_REG, bits,
-			      enable ? bits : 0, NULL);
+	ret = cci_update_bits(ov5693->regmap, OV5693_FORMAT1_REG, mask,
+			      enable ? ov5693_flip_vert_bits(ov5693) : 0, NULL);
 	if (ret)
 		return ret;
 
@@ -549,17 +594,73 @@ static const struct v4l2_ctrl_ops ov5693_ctrl_ops = {
 
 /* System Control Functions */
 
+/*
+ * Registers the vendor (Windows) driver programs differently for the full
+ * resolution and the 2x2 binned readout, taken from the mode tables in
+ * ov5693.sys.
+ *
+ * The mainline driver writes only the full-resolution values, once, from
+ * ov5693_global_regs, and never revisits them. Selecting the binned mode
+ * therefore leaves the sensor with a MIPI PLL multiplier (0x30b3) and
+ * analogue readout timing meant for twice the line length. On an IPU6 the
+ * receiver then reports "Transfer FIFO overflow" and every frame arrives
+ * empty, which is why the binned mode has been unusable -- and why the
+ * sensor has had to be read at full resolution and downscaled, costing two
+ * stops of light and forcing the AGC to its maximum analogue gain.
+ */
+static const struct cci_reg_sequence ov5693_mode_full_regs[] = {
+	{CCI_REG8(0x30b3), 0x83},	/* MIPI PLL multiplier */
+	{CCI_REG8(0x3620), 0x54},
+	{CCI_REG8(0x3621), 0xc7},
+	{CCI_REG8(0x3622), 0x0f},
+	{CCI_REG8(0x3708), 0xe2},
+	{CCI_REG8(0x3709), 0xc3},
+	{CCI_REG8(0x371f), 0x0c},
+};
+
+static const struct cci_reg_sequence ov5693_mode_binned_regs[] = {
+	{CCI_REG8(0x30b3), 0x70},	/* MIPI PLL multiplier */
+	{CCI_REG8(0x3600), 0xbc},	/* analogue, vendor sets it binned-only */
+	{CCI_REG8(0x3620), 0x44},
+	{CCI_REG8(0x3621), 0xb5},
+	{CCI_REG8(0x3622), 0x0c},
+	{CCI_REG8(0x3708), 0xe6},
+	{CCI_REG8(0x3709), 0xc7},
+	{CCI_REG8(0x371f), 0x1f},
+};
+
 static int ov5693_mode_configure(struct ov5693_device *ov5693)
 {
 	const struct ov5693_mode *mode = &ov5693->mode;
+	bool binned = mode->binning_x || mode->binning_y;
 	int ret = 0;
 
+	/*
+	 * Readout geometry. At full resolution this is the crop rectangle
+	 * selected over the media API, with the sensor-ISP window offsets at
+	 * zero.
+	 *
+	 * The binned mode instead copies the vendor driver: read the whole
+	 * pixel array (0..2623 x 0..1955), bin it to 1312x978, and let the
+	 * ISP window offsets place the 1296x972 output inside that. The
+	 * vertical offset is 2 rather than the vendor's 3: an odd offset
+	 * shifts the Bayer phase by one row (everything decodes magenta and
+	 * green); the vendor ISP knows the resulting phase, a V4L2 receiver
+	 * assumes it unchanged. The
+	 * obvious alternative -- keeping the 2592x1944 crop so the binned
+	 * readout exactly equals the output size with zero offset -- leaves
+	 * the sensor ISP without spare pixels around the window, and the
+	 * frames it emits then fail CSI-2 framing ("Frame sync error" on the
+	 * IPU6, no frame ever completes).
+	 */
+
 	/* Crop Start X */
-	cci_write(ov5693->regmap, OV5693_CROP_START_X_REG, mode->crop.left,
-		  &ret);
+	cci_write(ov5693->regmap, OV5693_CROP_START_X_REG,
+		  binned ? 0 : mode->crop.left, &ret);
 
 	/* Offset X */
-	cci_write(ov5693->regmap, OV5693_OFFSET_START_X_REG, 0, &ret);
+	cci_write(ov5693->regmap, OV5693_OFFSET_START_X_REG,
+		  binned ? 8 : 0, &ret);
 
 	/* Output Size X */
 	cci_write(ov5693->regmap, OV5693_OUTPUT_SIZE_X_REG, mode->format.width,
@@ -567,18 +668,20 @@ static int ov5693_mode_configure(struct ov5693_device *ov5693)
 
 	/* Crop End X */
 	cci_write(ov5693->regmap, OV5693_CROP_END_X_REG,
-		  mode->crop.left + mode->crop.width, &ret);
+		  binned ? OV5693_NATIVE_WIDTH - 1 :
+			   mode->crop.left + mode->crop.width, &ret);
 
 	/* Horizontal Total Size */
 	cci_write(ov5693->regmap, OV5693_TIMING_HTS_REG, OV5693_FIXED_PPL,
 		  &ret);
 
 	/* Crop Start Y */
-	cci_write(ov5693->regmap, OV5693_CROP_START_Y_REG, mode->crop.top,
-		  &ret);
+	cci_write(ov5693->regmap, OV5693_CROP_START_Y_REG,
+		  binned ? 0 : mode->crop.top, &ret);
 
-	/* Offset Y */
-	cci_write(ov5693->regmap, OV5693_OFFSET_START_Y_REG, 0, &ret);
+	/* Offset Y: parity picks the Bayer phase, see binned_y_offset. */
+	cci_write(ov5693->regmap, OV5693_OFFSET_START_Y_REG,
+		  binned ? binned_y_offset : 0, &ret);
 
 	/* Output Size Y */
 	cci_write(ov5693->regmap, OV5693_OUTPUT_SIZE_Y_REG, mode->format.height,
@@ -586,7 +689,8 @@ static int ov5693_mode_configure(struct ov5693_device *ov5693)
 
 	/* Crop End Y */
 	cci_write(ov5693->regmap, OV5693_CROP_END_Y_REG,
-		  mode->crop.top + mode->crop.height, &ret);
+		  binned ? OV5693_NATIVE_HEIGHT - 1 :
+			   mode->crop.top + mode->crop.height, &ret);
 
 	/* Subsample X increase */
 	cci_write(ov5693->regmap, OV5693_SUB_INC_X_REG,
@@ -604,6 +708,24 @@ static int ov5693_mode_configure(struct ov5693_device *ov5693)
 			OV5693_FORMAT2_HBIN_EN,
 			mode->binning_x ? OV5693_FORMAT2_HBIN_EN : 0, &ret);
 
+	/*
+	 * The set of FORMAT1 bits that means "vertical flip" changes with the
+	 * readout mode, so re-apply the control now that binning is known.
+	 */
+	if (ov5693->ctrls.vflip && ov5693->ctrls.vflip->val)
+		cci_update_bits(ov5693->regmap, OV5693_FORMAT1_REG,
+				OV5693_FORMAT1_FLIP_VERT_ISP_EN |
+				OV5693_FORMAT1_FLIP_VERT_SENSOR_EN,
+				ov5693_flip_vert_bits(ov5693), &ret);
+
+	/* PLL and analogue readout timing follow the readout mode. */
+	if (mode->binning_x || mode->binning_y)
+		cci_multi_reg_write(ov5693->regmap, ov5693_mode_binned_regs,
+				    ARRAY_SIZE(ov5693_mode_binned_regs), &ret);
+	else
+		cci_multi_reg_write(ov5693->regmap, ov5693_mode_full_regs,
+				    ARRAY_SIZE(ov5693_mode_full_regs), &ret);
+
 	return ret;
 }
 
@@ -611,6 +733,18 @@ static int ov5693_enable_streaming(struct ov5693_device *ov5693, bool enable)
 {
 	int ret = 0;
 
+	/*
+	 * Program MIPI_CTRL00 (CSI-2 clock-lane control) before starting the
+	 * stream. The 0x00 power-on default is accepted by the IPU3 ISP but
+	 * leaves the IPU6 D-PHY (Surface Pro 7+/8/9) unable to lock onto the
+	 * link: the sensor streams yet the receiver gets no CSI-2 data and
+	 * capture times out. 0x2d matches the value the vendor (Windows)
+	 * driver writes and lets the IPU6 receive frames.
+	 */
+	if (enable)
+		cci_write(ov5693->regmap, OV5693_MIPI_CTRL00_REG,
+			  OV5693_MIPI_CTRL00_IPU6, &ret);
+
 	cci_write(ov5693->regmap, OV5693_SW_STREAM_REG,
 		  enable ? OV5693_START_STREAMING : OV5693_STOP_STREAMING,
 		  &ret);
@@ -762,6 +896,16 @@ static unsigned int __ov5693_calc_vts(u32 height)
 
 	tgt_fps = rounddown(OV5693_PIXEL_RATE / OV5693_FIXED_PPL / height, 30);
 
+	/*
+	 * Cap at 30fps. Left alone this picks 60fps for the 2x2 binned mode,
+	 * which halves VTS to 1038 -- and the binned readout runs off a slower
+	 * MIPI PLL (0x30b3), so the link cannot carry twice the frame rate:
+	 * the IPU6 reports "Frame sync error" and no frame ever completes. The
+	 * vendor driver runs the same mode at VTS 1984. Staying at 30fps also
+	 * keeps the exposure ceiling, which is VTS bound, where it was.
+	 */
+	tgt_fps = min(tgt_fps, 30U);
+
 	return ALIGN_DOWN(OV5693_PIXEL_RATE / OV5693_FIXED_PPL / tgt_fps, 2);
 }
 
-- 
2.43.0

From 63db2dfcb98746ba6bc40c96945f736f29b42a1b Mon Sep 17 00:00:00 2001
From: "D. Manresa" <d.manresa@electronica-submarina.com>
Date: Thu, 27 Aug 2026 23:03:33 +0200
Subject: [PATCH] media: i2c: ov7251: drive the IR illuminator while streaming

On the Surface (Windows Hello) IR cameras the illuminator hangs off the
OV7251's strobe output and nothing in ACPI describes it: the INT3472
for this camera declares only a power-enable and a reset GPIO. Without
programming the strobe the sensor sees a completely black scene - the
module sits behind an IR-pass filter and room lighting emits next to
nothing in the infrared.

Two registers matter, both taken from the mode tables of the vendor's
ov7251.sys:

  0x3005  pad output enable; 0x08 routes the strobe to the illuminator.
  0x3b81  strobe pattern, one bit per frame. The vendor writes 0xff for
          its illuminated modes and 0xa5 for the rest. This driver's
          global init programs 0xa5 (10100101), so even with the pad
          enabled the illuminator fires on half the frames: the LED
          blinks visibly and every second frame comes out black.

Turn the illuminator on at stream-on and off at stream-off, so it is
lit exactly while an application is consuming frames. Enabled
automatically for ACPI id INT347E (the Surface discrete-INT3472 IR
module); a module parameter (ir_led) allows forcing it either way for
other designs where the wiring is unknown.

Verified on a Surface Pro 7+: IR frames are black without this, usable
for face enrolment/recognition with it.

Signed-off-by: D. Manresa <d.manresa@electronica-submarina.com>
Co-Authored-By: Claude Fable 5 <noreply@anthropic.com>
Patchset: cameras
---
 drivers/media/i2c/ov7251.c | 60 ++++++++++++++++++++++++++++++++++++++
 1 file changed, 60 insertions(+)

diff --git a/drivers/media/i2c/ov7251.c b/drivers/media/i2c/ov7251.c
index 28b192c..77e7183 100644
--- a/drivers/media/i2c/ov7251.c
+++ b/drivers/media/i2c/ov7251.c
@@ -23,6 +23,30 @@
 #include <media/v4l2-fwnode.h>
 #include <media/v4l2-subdev.h>
 
+/*
+ * The Windows-Hello IR illuminator hangs off the sensor's strobe output.
+ * Nothing in ACPI describes it -- the INT3472 for this camera declares only a
+ * power-enable and a reset GPIO -- so without programming it here the sensor
+ * sees a completely black scene: the module sits behind an IR-pass filter and
+ * room lighting emits next to nothing in the infrared.
+ *
+ * Two registers matter, both read out of the mode tables in the vendor's
+ * ov7251.sys:
+ *
+ *   0x3005  pad output value; 0x08 routes the strobe to the illuminator.
+ *   0x3b81  strobe pattern, one bit per frame. The vendor writes 0xff for its
+ *           illuminated modes and 0xa5 for the rest. 0xa5 is 10100101, so the
+ *           illuminator fires on half the frames: the LED blinks visibly and
+ *           every second frame comes out black. This driver's global init
+ *           programs 0xa5, hence the need to override it.
+ */
+#define OV7251_PAD_OUT1			0x3005
+#define OV7251_PAD_OUT1_IR_LED		0x08
+#define OV7251_STROBE_PATTERN		0x3b81
+#define OV7251_STROBE_EVERY_FRAME	0xff
+#define OV7251_STROBE_HALF_FRAMES	0xa5
+
+
 #define OV7251_SC_MODE_SELECT		0x0100
 #define OV7251_SC_MODE_SELECT_SW_STANDBY	0x0
 #define OV7251_SC_MODE_SELECT_STREAMING		0x1
@@ -1334,6 +1358,36 @@ static int ov7251_get_selection(struct v4l2_subdev *sd,
 	return 0;
 }
 
+static int ir_led = -1;
+module_param(ir_led, int, 0644);
+MODULE_PARM_DESC(ir_led,
+		 "IR illuminator while streaming: -1 auto (on for ACPI INT347E), 0 off, 1 on");
+
+static bool ov7251_has_ir_led(struct ov7251 *ov7251)
+{
+	if (ir_led >= 0)
+		return ir_led;
+
+	return acpi_dev_hid_uid_match(ACPI_COMPANION(ov7251->dev), "INT347E", NULL);
+}
+
+static int ov7251_ir_led_set(struct ov7251 *ov7251, bool on)
+{
+	int ret;
+
+	if (!ov7251_has_ir_led(ov7251))
+		return 0;
+
+	ret = ov7251_write_reg(ov7251, OV7251_PAD_OUT1,
+			       on ? OV7251_PAD_OUT1_IR_LED : 0);
+	if (ret)
+		return ret;
+
+	return ov7251_write_reg(ov7251, OV7251_STROBE_PATTERN,
+				on ? OV7251_STROBE_EVERY_FRAME :
+				     OV7251_STROBE_HALF_FRAMES);
+}
+
 static int ov7251_s_stream(struct v4l2_subdev *subdev, int enable)
 {
 	struct ov7251 *ov7251 = to_ov7251(subdev);
@@ -1372,7 +1426,13 @@ static int ov7251_s_stream(struct v4l2_subdev *subdev, int enable)
 				       OV7251_SC_MODE_SELECT_STREAMING);
 		if (ret)
 			goto err_power_down;
+
+		ret = ov7251_ir_led_set(ov7251, true);
+		if (ret)
+			goto err_power_down;
 	} else {
+		ov7251_ir_led_set(ov7251, false);
+
 		ret = ov7251_write_reg(ov7251, OV7251_SC_MODE_SELECT,
 				       OV7251_SC_MODE_SELECT_SW_STANDBY);
 		pm_runtime_put(ov7251->dev);
-- 
2.43.0

From fe6f21c96e7657d16a16c8df1eb82c7e9d46d19c Mon Sep 17 00:00:00 2001
From: "D. Manresa" <d.manresa@electronica-submarina.com>
Date: Thu, 27 Aug 2026 23:03:33 +0200
Subject: [PATCH] platform/x86: int3472: map the Surface Pro 7+ and Pro 8
 sensor rails

Three power-sequencing gaps keep the Surface Pro 7+ IR (OV7251,
INT347E) and rear (OV8865, INT347A) cameras and the Surface Pro 8
infrared camera (ST VD55G0, SMO55F0) from powering on; the affected
sensors fail their first I2C access with -EREMOTEIO (-121):

1. The INT347E declares its single gated rail as a power-enable GPIO,
   which is registered as an "avdd" regulator by default. The ov7251
   driver requests its supplies as vdda/vddd/vdddo (per the DT
   bindings), so it gets a dummy regulator and the sensor is never
   powered. Add an int3472_gpio_map[] entry renaming it to "vdda",
   like the existing INT347E reset-GPIO quirk above it.

2. The INT347A declares a second power rail with the vendor-specific
   GPIO type 0x08. The generic handling registers it under a "pwr1"
   name that no sensor driver consumes; on the Pro 7+ the OV8865 does
   not come up unless this rail is actually driven. Register it under
   a supply name the ov8865 driver requests, default "dvdd". Which of
   the driver's supplies it feeds on this board is not knowable from
   ACPI, so the name is a module parameter (ov8865_pwr1_con_id). The
   default "dvdd" is what the in-tree ov8865 driver requests; with
   GPIO_SUPPLY_NAME_LENGTH raised to 6 in this patchset, "dovdd" -
   the other candidate - fits as well.

3. The Surface Pro 8 infrared sensor (ST VD55G0, HID SMO55F0) declares
   its rails as GPIO types 0x10, 0x0b and 0x12, which the generic
   handling registers as "dovdd", "avdd" and "dvdd". The VD55G0
   driver requests its supplies as vcore, vddio and vana, so the
   regulators exist but are never claimed and the GPIOs are never
   driven. Add per-HID int3472_gpio_map[] entries renaming them
   (0x10 -> "vddio", 0x0b -> "vana", 0x12 -> "vcore"); the map
   lookup runs before the generic type switch, so these win.
   Supplies verified by regulator consumer binding on a Surface
   Pro 8 by @fildunsky (linux-surface PR #2252).

Verified on a Surface Pro 7+: with both rails mapped the OV7251 and
OV8865 probe and stream; without either one the respective sensor
fails probe with -121. The SMO55F0 mappings were verified on a
Surface Pro 8 by the reporter of PR #2252 (regulator claimed by
i2c-SMO55F0:00, rail driven, illuminated frames captured).

Signed-off-by: D. Manresa <d.manresa@electronica-submarina.com>
Co-Authored-By: Claude Fable 5 <noreply@anthropic.com>
Patchset: cameras
---
 drivers/platform/x86/intel/int3472/discrete.c | 62 +++++++++++++++++++
 1 file changed, 62 insertions(+)

diff --git a/drivers/platform/x86/intel/int3472/discrete.c b/drivers/platform/x86/intel/int3472/discrete.c
index f0b72b1..cc9c4d3 100644
--- a/drivers/platform/x86/intel/int3472/discrete.c
+++ b/drivers/platform/x86/intel/int3472/discrete.c
@@ -157,6 +157,47 @@ static const struct int3472_gpio_map int3472_gpio_map[] = {
 		.type_to = INT3472_GPIO_TYPE_RESET,
 		.con_id = "enable",
 	},
+	{	/*
+		 * Surface Pro 7+ (and other IPU6 designs) wire the OV7251's
+		 * single gated rail to the INT3472 power-enable GPIO, but the
+		 * ov7251 driver / DT-bindings name their supplies vdda, vddd
+		 * and vdddo. Without this the regulator is registered as
+		 * "avdd", the driver finds no supply at all, and the sensor is
+		 * never powered: probe dies on the first I2C write with -121.
+		 */
+		.hid = "INT347E",
+		.type_from = INT3472_GPIO_TYPE_POWER_ENABLE,
+		.type_to = INT3472_GPIO_TYPE_POWER_ENABLE,
+		.con_id = "vdda",
+		.enable_time_us = GPIO_REGULATOR_ENABLE_TIME,
+	},
+	{	/*
+		 * The Surface Pro 8's infrared sensor (ST VD55G0, SMO55F0)
+		 * requests its supplies as vcore, vddio and vana, so the
+		 * generic "dovdd"/"avdd"/"dvdd" rails are never claimed and
+		 * probe dies with -121. Supplies verified by regulator
+		 * consumer binding on Surface Pro 8, see PR #2252.
+		 */
+		.hid = "SMO55F0",
+		.type_from = 0x10,
+		.type_to = 0x10,
+		.con_id = "vddio",
+		.enable_time_us = GPIO_REGULATOR_ENABLE_TIME,
+	},
+	{
+		.hid = "SMO55F0",
+		.type_from = INT3472_GPIO_TYPE_POWER_ENABLE,
+		.type_to = INT3472_GPIO_TYPE_POWER_ENABLE,
+		.con_id = "vana",
+		.enable_time_us = GPIO_REGULATOR_ENABLE_TIME,
+	},
+	{
+		.hid = "SMO55F0",
+		.type_from = INT3472_GPIO_TYPE_HANDSHAKE,
+		.type_to = INT3472_GPIO_TYPE_HANDSHAKE,
+		.con_id = "vcore",
+		.enable_time_us = GPIO_REGULATOR_ENABLE_TIME,
+	},
 	{	/* ov08x40's handshake pin needs a 45 ms delay on some HP laptops */
 		.hid = "OVTI08F4",
 		.type_from = INT3472_GPIO_TYPE_HANDSHAKE,
@@ -166,6 +207,19 @@ static const struct int3472_gpio_map int3472_gpio_map[] = {
 	},
 };
 
+/*
+ * Surface Pro 7+ declares a second power-enable GPIO for the OV8865 with the
+ * vendor-specific type 0x08, which no upstream code handles. The sensor needs
+ * it: with only the type 0x0B rail ("avdd") powered the first I2C write dies
+ * with -121. Which of the ov8865 driver's supplies it actually feeds is a
+ * guess, so keep it settable. The default "dvdd" is what the in-tree ov8865
+ * driver requests; GPIO_SUPPLY_NAME_LENGTH is 6 here, so "dovdd" fits too.
+ */
+static char *ov8865_pwr1_con_id = "dvdd";
+module_param(ov8865_pwr1_con_id, charp, 0644);
+MODULE_PARM_DESC(ov8865_pwr1_con_id,
+		 "supply name for the OV8865's type 0x08 power-enable GPIO");
+
 static void int3472_get_con_id_and_polarity(struct int3472_discrete_device *int3472, u8 *type,
 					    const char **con_id, unsigned long *gpio_flags,
 					    unsigned int *enable_time_us)
@@ -173,6 +227,14 @@ static void int3472_get_con_id_and_polarity(struct int3472_discrete_device *int3
 	struct acpi_device *adev = int3472->sensor;
 	unsigned int i;
 
+	if (*type == 0x08 && acpi_dev_hid_uid_match(adev, "INT347A", NULL)) {
+		*type = INT3472_GPIO_TYPE_POWER_ENABLE;
+		*con_id = ov8865_pwr1_con_id;
+		*gpio_flags = GPIO_ACTIVE_HIGH;
+		*enable_time_us = GPIO_REGULATOR_ENABLE_TIME;
+		return;
+	}
+
 	for (i = 0; i < ARRAY_SIZE(int3472_gpio_map); i++) {
 		/*
 		 * Map the firmware-provided GPIO to whatever a driver expects
-- 
2.43.0

